rskj-core/src/main/java/org/ethereum/datasource/RocksDbDataSource.java

```diff
@@ -0,0 +1,364 @@
+ * (derived from ethereumJ library, Copyright (c) 2016 <ether.camp>)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.ethereum.datasource;
+
+import co.rsk.metrics.profilers.Metric;
+import co.rsk.metrics.profilers.Profiler;
+import co.rsk.metrics.profilers.ProfilerFactory;
+import co.rsk.panic.PanicProcessor;
+import org.ethereum.db.ByteArrayWrapper;
+import org.ethereum.util.ByteUtil;
+import org.rocksdb.*;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.*;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import static java.lang.System.getProperty;
+
+public class RocksDbDataSource implements KeyValueDataSource {
+
+    private static final Logger logger = LoggerFactory.getLogger("db");
+    private static final Profiler profiler = ProfilerFactory.getInstance();
+    private static final PanicProcessor panicProcessor = new PanicProcessor();
+
+    private final String databaseDir;
+    private final String name;
+    private RocksDB db;
+    private boolean alive;
+
+    // The native LevelDB insert/update/delete are normally thread-safe
+    // However close operation is not thread-safe and may lead to a native crash when
+    // accessing a closed DB.
+    // The rocksdbJNI lib has a protection over accessing closed DB but it is not synchronized
+    // This ReadWriteLock still permits concurrent execution of insert/delete/update operations
+    // however blocks them on init/close/delete operations
+    private final ReadWriteLock resetDbLock = new ReentrantReadWriteLock();
+
+    public RocksDbDataSource(String name, String databaseDir) {
+        this.databaseDir = databaseDir;
+        this.name = name;
+        logger.debug("New RocksDbDataSource: {}", name);
+    }
+
+    public static KeyValueDataSource makeDataSource(Path datasourcePath) {
+        KeyValueDataSource ds = new RocksDbDataSource(datasourcePath.getFileName().toString(), datasourcePath.getParent().toString());
+        ds.init();
+        return ds;
+    }
+
+    @Override
+    public void init() {
+        resetDbLock.writeLock().lock();
+        Metric metric = profiler.start(Profiler.PROFILING_TYPE.LEVEL_DB_INIT);
```

Comment:
Its interesting that we create a new constant name for the profiles to diferentiate it from the LevelDB

Reply:
Done